ecl/hthor/hthor.cpp

```diff
@@ -1730,24 +1759,38 @@ class CHThorPipeWriteActivity : public CHThorActivityBase
 
     virtual void execute()
     {
-        for (;;)
+        Owned<IPipeProcessException> pipeException;
+        try
         {
-            const void *row = input->nextRow();
-            if (!row)
+            for (;;)
             {
-                row = input->nextRow();
+                OwnedConstRoxieRow row(input->nextRow());
                 if (!row)
-                    break;
+                {
+                    row.setown(input->nextRow());
+                    if (!row)
+                        break;
+                }
+                processed++;
+                if (recreate)
+                    openPipe(helper.getNameFromRow(row));
+                writeTransformer->writeTranslatedText(row, pipe);
+                if (recreate)
+                {
+                    closePipe();
+                    verifyPipe();
+                }
             }
-            processed++;
-            if(recreate)
-                openPipe(helper.getNameFromRow(row));
-            writeTransformer->writeTranslatedText(row, pipe);
-            ReleaseRoxieRow(row);
-            if(recreate)
+            if (!recreate)
```

Comment:
!recreate or recreate ?

Reply:
or should this closePipe() block be here ?

Reply:
if (!recreate) is correct.
This is outside of the for loop, i.e. it is finishing. If it is not a pipe that is recreated per row, it should close it now.
The way the code was before:
```
        for (;;)
        {
...
..
            if(recreate)
                closePipe();
        }
        closePipe();
```

.. it used to incorrectly call closePipe() twice in the !recreate case, once for end of row and once after for loop. It should only have called it, if not already closed, i.e. if !recreate